### PR TITLE
Notifications support

### PR DESCRIPTION
**Pull request by nicolasmorini**
Hi @rjarry!

Please consider this pull request that adds support for notifications.

This pull-request depends directly on the following pull-request on libyang-python: https://github.com/CESNET/libyang-python/pull/26

This is the follow up of the discussion here: sysrepo/sysrepo-python#10 (apologies for the delay on making this contribution).

Please note that I had to do several changes in the CI workflow to make it build and pass, since the release of libyang2, new versions of pylint and ubuntu-latest resolving to ubuntu-20.04 broke the process.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -154,7 +154,7 @@ def make_release_tree(self, base_dir, files):
         'cffi; platform_python_implementation != "PyPy"',
     ],
     install_requires=[
-        "libyang>=1.5,<2",
+        "libyang>1.6.1,<2",
```
Comment (nicolasmorini) [inline setup.py:157]: This probably needs to be adapted accordingly once the corresponding libyang-python release is available. For now I just wanted to signal that it requires a version that does not exist yet.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -37,10 +37,12 @@ jobs:
             toxenv: py38
           - python: pypy3
             toxenv: pypy3
+          #- python: 3.x
+          #  toxenv: py3-master
+          #- python: 3.x
+          #  toxenv: py3-devel
```
Comment (rjarry) [inline .github/workflows/ci.yml:43]: I guess you can remove the master and devel branches. I'll add them again when working on libyang 2 support.

File: cffi/build.py [hunk 3]
```diff
@@ -11,7 +11,7 @@
 HERE = os.path.dirname(__file__)
 
 BUILDER = cffi.FFI()
-with open(os.path.join(HERE, "cdefs.h")) as f:
+with open(os.path.join(HERE, "cdefs.h"), encoding="utf-8") as f:
```
Comment (rjarry) [inline cffi/build.py:14]: Did you have a problem with the default encoding? These files only contain ASCII characters as far as I know.

File: tox.ini [hunk 4]
```diff
@@ -2,7 +2,7 @@
 # SPDX-License-Identifier: BSD-3-Clause
 
 [tox]
-envlist = format,lint,py{35,36,37,38,py3},py3-{master,devel},coverage
+envlist = format,lint,py{35,36,37,38,py3},py3-{master,devel,libyang1},coverage
```
Comment (rjarry) [inline tox.ini:5]: You can remove the master and devel branches.

File: tox.ini [hunk 5]
```diff
@@ -21,6 +21,8 @@ setenv =
 	master: SYSREPO_BRANCH = master
 	devel: LIBYANG_BRANCH = devel
 	devel: SYSREPO_BRANCH = devel
```
Comment (rjarry) [inline tox.ini:23]: Same here about the branches.

File: setup.py (hunk 1 above)
Comment (rjarry) [reply setup.py:157]: Ok, I'll integrate your work on https://github.com/CESNET/libyang-python/pull/26 first and make an official release. Then you'll be able to update your code with the new version.

File: sysrepo/subscription.py [hunk 6]
```diff
@@ -487,3 +495,45 @@ def rpc_callback(session, xpath, input_node, event, req_id, output_node, priv):
         if isinstance(session, SysrepoSession) and isinstance(xpath, str):
             session.set_error(xpath, str(e))
         return lib.SR_ERR_CALLBACK_FAILED
+
+
```
Comment (rjarry) [inline sysrepo/subscription.py:499]: nitpick: could you add a separator line? I am quite anal about these :)

```python
# ------------------------------------------------------------------------------
```

File: sysrepo/subscription.py [hunk 7]
```diff
@@ -487,3 +495,45 @@ def rpc_callback(session, xpath, input_node, event, req_id, output_node, priv):
...
+    private_data = subscription.private_data
+    notif_type = NOTIF_TYPES[notif_type]
+
+    ly_ctx = session.get_ly_ctx()
+    notif_dnode = DNode.new(ly_ctx, notif)
+    xpath = notif_dnode.path()
+    notif_dict = notif_dnode.print_dict(
+        include_implicit_defaults=subscription.include_implicit_defaults
+    )
+    callback(xpath, notif_type, notif_dict, timestamp, private_data)
```
Comment (rjarry) [inline sysrepo/subscription.py:539]: You should wrap all this in a try/except block catching all exceptions. See all C callback handlers for reference.

File: sysrepo/session.py [hunk 8]
```diff
@@ -450,6 +450,107 @@ def subscribe_rpc_call(
...
+                type uint32;
+              }
+              leaf param2 {
+                type string;
+              }
+            }
+
+        The notification dict may look like this::
+
+            {'some-notification': {'param1': 42, 'param2': 'foobar'}}
```
Comment (rjarry) [inline sysrepo/session.py:477]: I feel like the notification dict should look like the one in RPCs and actions. See commit https://github.com/sysrepo/sysrepo-python/commit/3073a2cfdb8662e65645a63ce628ad3ad4706b4c for the reasoning around this design decision. Basically, the xpath allows identifying the data-tree position of the rpc/action and the dict contains only the input/output nodes.

https://github.com/sysrepo/sysrepo-python/blob/master/sysrepo/session.py#L335-L365

https://github.com/sysrepo/sysrepo-python/blob/master/sysrepo/subscription.py#L465-L469

File: sysrepo/session.py [hunk 9]
```diff
@@ -1000,6 +1101,52 @@ def rpc_send(
...
+        :raises SysrepoError:
+            If the notification callback failed.
+        """
+
+        ctx = self.get_ly_ctx()
+        module_name, _, _ = next(libyang.xpath_split(xpath))
+        module = ctx.get_module(module_name)
+        dnode = module.parse_data_dict(
+            notification, notification=True, strict=strict, validate=False
+        )
```
Comment (rjarry) [inline sysrepo/session.py:1144]: Following my previous comment, the notification dict should only contain the inner nodes and the xpath should be used to construct the complete data tree. See commit https://github.com/sysrepo/sysrepo-python/commit/1a070313ee4e9dec5b30c5f4b0269a2cf8dabaf0.

https://github.com/sysrepo/sysrepo-python/blob/master/sysrepo/session.py#L981-L984

File: tests/test_subs_notification.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        # reconnect to make sure module is removed
+        with sysrepo.SysrepoConnection(err_on_sched_fail=True):
+            pass
+
+    def test_notification_sub(self):
+        priv = object()
+        notif_xpath = "/sysrepo-example:alarm-triggered"
+        notif_dict = {
+            "alarm-triggered": {"description": "An error occurred", "severity": 3}
+        }
```
Comment (rjarry) [inline tests/test_subs_notification.py:39]: After changing the notification dict structure, could you add a test with a notification [nested in a data element](https://datatracker.ietf.org/doc/html/rfc7950#section-7.16.3)? You can have a look at commit https://github.com/sysrepo/sysrepo-python/commit/1a070313ee4e9dec5b30c5f4b0269a2cf8dabaf0 for inspiration.

**Review by rjarry — CHANGES_REQUESTED** [review]
Hi @nicolasmorini,

overall this looks good. Could you address the few remarks I have made?

Thanks.

File: cffi/build.py (hunk 3 above)
Comment (nicolasmorini) [reply cffi/build.py:14]: I added it to make pylint happy (it is a new warning added recently):

- unspecified-encoding (W1514) (https://pylint.pycqa.org/en/latest/technical_reference/features.html)
- https://pylint.pycqa.org/en/latest/whatsnew/2.10.html

Alternatively, I can disable that specific warning, either inline or altogether in pylintrc. What are your thoughts on this?

**Comment by nicolasmorini** [comment]
@rjarry, I'll address the remaining comments about dict usage and exception handling later today

File: cffi/build.py (hunk 3 above)
Comment (rjarry) [reply cffi/build.py:14]: Ok that looks good then, leave it like this :+1:

**Comment by rjarry** [comment]
[Libyang 1.7.0 is out](https://pypi.org/project/libyang/). You can update `setup.py` accordingly :)

File: setup.py (hunk 1 above)
Comment (rjarry) [reply setup.py:157]: [Libyang 1.7.0](https://pypi.org/project/libyang/) is out :)

**Comment by rjarry** [comment]
@nicolasmorini could you squash all your fixups please? You can ignore the codecov report. It is impossible to have coverage on C callbacks...

**Comment by rjarry** [comment]
Included in [0.5.0](https://pypi.org/project/sysrepo/), thanks @nicolasmorini :+1:

**Comment by nicolasmorini** [comment]
Thanks a lot for the support and the prompt responses @rjarry! I will start using these new releases this week!

**Pull request merged**